ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/resources/jts/orbspecific/XAResourceRecord.java

```diff
@@ -397,27 +397,25 @@ public void rollback() throws org.omg.CORBA.SystemException,
 						case XAException.XA_RBTIMEOUT:
 							destroyState();
 							break;
-						case XAException.XAER_NOTA:
-							if (_phaseTwoStarted) {
-								// rolled back previously and recovery completed
-								destroyState();
-								break;
-							}
-						case XAException.XAER_RMFAIL:
-							if (_prepared) {
-								// We do not want to delete the state on disk because that would prevent the recovery
-								// manager from reattempting recovery and XAER_RMFAIL out of xa_rollback is transient
-								// failure
-								destroyState = false;
-							}
-							break;
-						default:
-							destroyState();
-
-							if (_prepared)
-								throw new org.omg.CosTransactions.HeuristicHazard();
-							else
-								throw new UNKNOWN();
+                        case XAException.XAER_NOTA:
+                            if (_phaseTwoStarted) {
+                                // rolled back previously and recovery completed
+                                destroyState();
+                                break;
+                            }
+                            destroyState();
```

Comment:
The only change introduced by the PR are these 5 lines which have been moved from the default arm into the XAER_NOTA arm of the switch (the reformatting introduced in the PR made that difficult to spot and the default are deleted).

But I think the default arm needs be present (see my previous comment) and, if you agree, then you just need to let the `case XAException.XAER_NOTA` arm of the switch fall through into the default arm, ie the full patch would be:

```
						case XAException.XAER_NOTA:
							if (_phaseTwoStarted) {
								// rolled back previously and recovery completed
								destroyState();
								break;
							}
							// FALL THROUGH
						default:
                                                        etc ...
```